File: src/main/java/gargoyle/ct/ui/impl/control/CTLocalizableAction.java
package gargoyle.ct.ui.impl.control;

import gargoyle.ct.messages.MessageProvider;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;

@SuppressWarnings("ConstructorNotProtectedInAbstractClass")
public abstract class CTLocalizableAction extends CTAction {
    private static final String KEY_TEXT = "text-key";
    private static final String KEY_TOOLTIP = "tooltip-key";
    private final MessageProvider messages;

    public CTLocalizableAction(MessageProvider messages, String textKey) {
        this.messages = messages;
        setTextKey(textKey);
        setToolTipTextKey(textKey);
        setEnabled(true);
    }

    public CTLocalizableAction(MessageProvider messages, String textKey, Icon icon) {
        this.messages = messages;
        setTextKey(textKey);
        setToolTipTextKey(textKey);
        setIcon(icon);
        setEnabled(true);
    }

    public CTLocalizableAction(MessageProvider messages, String textKey, String tooltipTextKey, Icon icon) {
        this.messages = messages;
        setTextKey(textKey);
        setToolTipTextKey(tooltipTextKey);
        setIcon(icon);
        setEnabled(true);
    }

    protected CTLocalizableAction(MessageProvider messages, String textKey, String tooltipTextKey) {
        this.messages = messages;
        setTextKey(textKey);
        setToolTipTextKey(tooltipTextKey);
        setEnabled(true);
    }

    @Override
    public void init(@NotNull AbstractButton menuItem) {
        super.init(menuItem);
        menuItem.setText(getLocalizedText());
        menuItem.setToolTipText(getToolTipLocalizedText());
    }

    @NotNull
    protected final String getLocalizedText() {
        String textKey = getTextKey();
        return textKey == null || textKey.isEmpty() ? "" : messages.getMessage(textKey);
    }

    private String getTextKey() {
        return String.valueOf(getValue(KEY_TEXT));
    }

    private void setTextKey(String textKey) {
        putValue(KEY_TEXT, textKey);
    }

    @NotNull
    protected final String getToolTipLocalizedText() {
        String toolTipTextKey = getToolTipTextKey();
        return toolTipTextKey == null || toolTipTextKey.isEmpty() ? "" : messages.getMessage(toolTipTextKey);
    }

    private String getToolTipTextKey() {
        return String.valueOf(getValue(KEY_TOOLTIP));
    }

    private void setToolTipTextKey(String toolTipTextKey) {
        putValue(KEY_TOOLTIP, toolTipTextKey);
    }
}

File: src/main/kotlin/gargoyle/ct/ui/impl/control/CTLocalizableAction.kt
package gargoyle.ct.ui.impl.control

import gargoyle.ct.messages.MessageProvider
import javax.swing.AbstractButton
import javax.swing.Icon

abstract class CTLocalizableAction : CTAction {
    private val messages: MessageProvider

    constructor(messages: MessageProvider, textKey: String) {
        this.messages = messages
        this.textKey = textKey
        toolTipTextKey = textKey
        isEnabled = true
    }

    constructor(messages: MessageProvider, textKey: String, icon: Icon?) {
        this.messages = messages
        this.textKey = textKey
        toolTipTextKey = textKey
        setIcon(icon!!)
        isEnabled = true
    }

    constructor(messages: MessageProvider, textKey: String, tooltipTextKey: String, icon: Icon?) {
        this.messages = messages
        this.textKey = textKey
        toolTipTextKey = tooltipTextKey
        setIcon(icon!!)
        isEnabled = true
    }

    protected constructor(messages: MessageProvider, textKey: String, tooltipTextKey: String) {
        this.messages = messages
        this.textKey = textKey
        toolTipTextKey = tooltipTextKey
        isEnabled = true
    }

    override fun init(menuItem: AbstractButton) {
        super.init(menuItem)
        menuItem.text = localizedText
        menuItem.toolTipText = toolTipLocalizedText
    }

    val localizedText: String
        get() {
            val textKey: String = textKey
            return if (textKey.isEmpty()) "" else messages.getMessage(textKey)
        }
    private var textKey: String
        get() = getValue(KEY_TEXT).toString()
        private set(textKey) {
            putValue(KEY_TEXT, textKey)
        }
    val toolTipLocalizedText: String
        get() {
            val toolTipTextKey: String = toolTipTextKey
            return if (toolTipTextKey.isEmpty()) "" else messages.getMessage(toolTipTextKey)
        }
    private var toolTipTextKey: String
        get() = getValue(KEY_TOOLTIP).toString()
        private set(toolTipTextKey) {
            putValue(KEY_TOOLTIP, toolTipTextKey)
        }

    companion object {
        private const val KEY_TEXT = "text-key"
        private const val KEY_TOOLTIP = "tooltip-key"
    }
}
